File: lista9/trafficnetwork.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <limits>
using namespace std;
const long long MAX_LENGTH = 1000000000; // equivale a "infinito" 
typedef unordered_map<int, unordered_map<int, int>> AdjacencyMap;
typedef priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> MinPriorityQueue;

vector<int> dijkstra(AdjacencyMap& adj, int n, int origin, MinPriorityQueue& heap) {
    vector<int> distance(n + 1, MAX_LENGTH);
    distance[origin] = 0;
    while (!heap.empty()) {
        int dist = heap.top().first;
        int u = heap.top().second;
        heap.pop();

        if (dist > distance[u]) {
            continue;
        }

        for (auto& kv : adj[u]) {
            int v = kv.first;
            int w = kv.second;
            int oldDist = distance[v];
            int newDist = distance[u] + w;

            if (oldDist > newDist) {
                distance[v] = newDist;
                heap.push({newDist, v});
            }
        }
    }
    return distance;
}

tuple<AdjacencyMap, AdjacencyMap, MinPriorityQueue, MinPriorityQueue> initialize(int n, int origin, int end) {
    AdjacencyMap adj, adjReverse;
    MinPriorityQueue heap, heapReverse;
    long long lengthOrigin;
    long long lengthEnd;

    for (int i = 1; i <= n; i++) {
        adj[i] = {};
        adjReverse[i] = {};
        if(i == origin) {
            lengthOrigin = 0;
        } else {
            lengthOrigin = MAX_LENGTH;
        }
        if(i == end) {
            lengthEnd = 0;
        } else {
            lengthEnd = MAX_LENGTH;
        }
        heap.push({lengthOrigin, i});
        heapReverse.push({lengthEnd, i});
    }
    return { adj, adjReverse, heap, heapReverse };
}
int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int d;
    cin >> d;
    vector<int> paths;
    for (int i = 0; i < d; i++) {
        int n, m, k, s, t;
        cin >> n >> m >> k >> s >> t;
        auto [adj, adjReverse, heap, heapReverse] = initialize(n, s, t);

        for (int j = 0; j < m; j++) {
            int u, v, w;
            cin >> u >> v >> w;
            adj[u][v] = w;
            adjReverse[v][u] = w;
        }
        vector<int> distance = dijkstra(adj, n, s, heap);
        vector<int> distanceReverse = dijkstra(adjReverse, n, t, heapReverse);
        int pathLength = distance[t];
        for (int j = 0; j < k; j++) {
            int u, v, w;
            cin >> u >> v >> w;
            int newDistDirect = distance[u] + w + distanceReverse[v];
            int newDistReverse = distance[v] + w + distanceReverse[u];
            int newDist = min(newDistDirect, newDistReverse);

            if (newDist < pathLength && newDist < MAX_LENGTH) {
                pathLength = newDist;
            }
        }
        paths.push_back(pathLength);
    }
    for (long long i : paths) {
        if(i >= MAX_LENGTH) { i = -1;}
        cout << i << "\n";
    }
    return 0;
}

File: lista9/trafficnetwork.py
import heapq

def djikstra(adj,n,origin,heap):
    distance = [float('inf')] * (n+1)
    distance[origin] = 0
    while len(heap) > 0:
        dist,u = heapq.heappop(heap)
        if dist > distance[u]:
            continue
        for v in adj[u]:
            w = adj[u][v]
            oldDist = distance[v]
            newDist = distance[u] + w
            if oldDist > newDist:
                distance[v] = newDist
                heapq.heappush(heap,(newDist,v))
    return distance

def initialize(n,origin,end):
    adj = {}
    adjReverse = {}
    heap = []
    heapReverse = []
    for i in range(1,n+1):
        adj[i] = {}
        adjReverse[i] = {}
        heap.append((float('inf'),i))
        heapReverse.append((float('inf'),i))
    heap[origin-1] = (0,origin)
    heapReverse[end-1] = (0,end)
    return adj,adjReverse,heap,heapReverse

d = int(input())
paths = []
for _ in range(d):
    n,m,k,s,t = map(int,input().split())
    adj,adjReverse,heap,heapReverse = initialize(n,s,t)
    for _ in range(m):
        u,v,w = map(int,input().split())
        adj[u][v] = w
        adjReverse[v][u] = w
    distance = djikstra(adj,n,s,heap)
    distanceReverse = djikstra(adjReverse,n,t,heapReverse)
    pathLength = distance[t]
    for _ in range(k):
        u,v,w = map(int,input().split())
        newDistDirect = distance[u] + w + distanceReverse[v]
        newDistReverse = distance[v] + w + distanceReverse[u]
        newDist = min(newDistDirect,newDistReverse)
        if newDist < pathLength:
            pathLength = newDist
    paths.append(pathLength)
for p in paths:
    if p == float('inf'):
        print(-1)
    else:
        print(p)